File: software/examples/rust/example_threshold.rs
use std::{error::Error, io, thread};
use tinkerforge::{ip_connection::IpConnection, sound_intensity_bricklet::*};

const HOST: &str = "localhost";
const PORT: u16 = 4223;
const UID: &str = "XYZ"; // Change XYZ to the UID of your Sound Intensity Bricklet.

fn main() -> Result<(), Box<dyn Error>> {
    let ipcon = IpConnection::new(); // Create IP connection.
    let si = SoundIntensityBricklet::new(UID, &ipcon); // Create device object.

    ipcon.connect((HOST, PORT)).recv()??; // Connect to brickd.
                                          // Don't use device before ipcon is connected.

    // Get threshold receivers with a debounce time of 1 second (1000ms).
    si.set_debounce_period(1000);

    // Create receiver for intensity reached events.
    let intensity_reached_receiver = si.get_intensity_reached_receiver();

    // Spawn thread to handle received events. This thread ends when the `si` object
    // is dropped, so there is no need for manual cleanup.
    thread::spawn(move || {
        for intensity_reached in intensity_reached_receiver {
            println!("Intensity: {}", intensity_reached);
        }
    });

    // Configure threshold for intensity "greater than 2000".
    si.set_intensity_callback_threshold('>', 2000, 0);

    println!("Press enter to exit.");
    let mut _input = String::new();
    io::stdin().read_line(&mut _input)?;
    ipcon.disconnect();
    Ok(())
}

File: software/examples/go/example_threshold.go
package main

import (
	"fmt"
	"github.com/tinkerforge/go-api-bindings/ipconnection"
	"github.com/tinkerforge/go-api-bindings/sound_intensity_bricklet"
)

const ADDR string = "localhost:4223"
const UID string = "XYZ" // Change XYZ to the UID of your Sound Intensity Bricklet.

func main() {
	ipcon := ipconnection.New()
	defer ipcon.Close()
	si, _ := sound_intensity_bricklet.New(UID, &ipcon) // Create device object.

	ipcon.Connect(ADDR) // Connect to brickd.
	defer ipcon.Disconnect()
	// Don't use device before ipcon is connected.

	// Get threshold receivers with a debounce time of 1 second (1000ms).
	si.SetDebouncePeriod(1000)

	si.RegisterIntensityReachedCallback(func(intensity uint16) {
		fmt.Printf("Intensity: %d\n", intensity)
	})

	// Configure threshold for intensity "greater than 2000".
	si.SetIntensityCallbackThreshold('>', 2000, 0)

	fmt.Print("Press enter to exit.")
	fmt.Scanln()

}
